### PR TITLE
PPO training loop

### PR DESCRIPTION
**Pull request by nviolante25**
# PPO training loop

[x] Rollout phase
[x] General Advantage estimation
[x] Batched training

### CODE REVIEW

File: ppo/agents/base_agent.py [hunk 1]
```diff
@@ -53,17 +51,26 @@ def policy(policy_params: hk.Params, observation: np.ndarray):
...
         # Convert to get a batch shape
-        observation = tree_util.tree_map(
-            lambda x: jnp.expand_dims(x, axis=0), observation
-        )
+        observation = tree_util.tree_map(lambda x: jnp.expand_dims(x, axis=0), observation)
 
-        action = self.sampling_policy(
-            self.policy_params, next(self.sampling_keys), observation
-        )
+        action, log_prob = self.sampling_policy(self.policy_params, next(self.greedy_keys), observation)
```
Comment (theovincent) [inline ppo/agents/base_agent.py:58]: This needs to be changed :)
```suggestion
        action, log_prob = self.sampling_policy(self.policy_params, next(self.sampling_keys), observation)
```

File: ppo/agents/advantage.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            not_done_tp1 = 1.0 - trajectory[t].done_tp1
+            value_tp1 = trajectory[t + 1].value_t
+        delta = trajectory[t].reward_tp1 + discount * value_tp1 * not_done_tp1 - trajectory[t].value_t
+        advantages = advantages.at[t].set((delta + discount * gae_lambda * not_done_tp1 * lastgaelam)[0])
+        lastgaelam = advantages[t]
+
+    return advantages
+
+
+def add_advantage(advantages, agent):
```
Comment (theovincent) [inline ppo/agents/advantage.py:36]: It might be "cleaner" to add a method to `VanillaPPO`.
This method would call another method of `BaseReplayBuffer` adding the advantage to the memory.

This means that I would suggest adding two methods, one in `VanillaPPO` and one in `BaseReplayBuffer`.

In the main loop it would look like:
```Python
agent.add_advantage(advantage)
```

**Review by theovincent — CHANGES_REQUESTED** [review]

**Review by theovincent — APPROVED** [review]
